### PR TITLE
[Fix #7299] Implement Lint/RaiseException cop

### LINKED ISSUE #7299: alert on `raise Exception` usage
similar to the "do not rescue Exception" cop it would make sense to not allow raising Exception since that will bypass test frameworks and error reporters

### PR DESCRIPTION
**Pull request by denys281**
Resolves https://github.com/rubocop-hq/rubocop/issues/7299
Should go with https://github.com/rubocop-hq/ruby-style-guide/pull/775

Cop checks for:
- `raise Exception`;
- `raise Exception, 'error message'`;
- `raise Exception.new('Error message')
-  Also it works for `fail` and `::Exception`

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](https://github.com/rubocop-hq/rubocop/blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop-hq/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Run `bundle exec rake default`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: lib/rubocop/cop/lint/raise_exception.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+        def_node_matcher :exception_new_with_message?, <<~PATTERN
+          (send nil? ${:raise :fail}
+            (send (const _ :Exception) :new ... ))
+        PATTERN
+
+        def on_send(node)
+          offense = exception?(node) || exception_new_with_message?(node)
+
+          add_offense(node, message: format(MSG, method: offense)) if offense
```
Comment (Drenmi) [inline lib/rubocop/cop/lint/raise_exception.rb]: You could combine the matchers and do:

```ruby
raise_exception?(node) do |method|
  add_offense(node, message: format(MSG, method: method))
end
```

**Comment by Drenmi** [comment]
Thanks @denys281! 🙏

File: lib/rubocop/cop/lint/raise_exception.rb (hunk 1 above)
Comment (denys281) [reply lib/rubocop/cop/lint/raise_exception.rb]: And how to combine them? I tried, but without success. You meant to write universal matcher or to combine two int one? 

or you meant  ? :) 

```
def raise_exception(node)
  exception?(node) || exception_new_with_message?(node)
end
```

File: config/default.yml [hunk 2]
```diff
@@ -1505,6 +1505,11 @@ Lint/PercentSymbolArray:
   Enabled: true
   VersionAdded: '0.41'
 
+Lint/RaiseException:
+  Description: Checks for `raise` or `fail` statements which is raise `Exception` class.
```
Comment (bbatsov) [inline config/default.yml]: which are raising

File: config/default.yml [hunk 3]
```diff
@@ -1505,6 +1505,11 @@ Lint/PercentSymbolArray:
...
+  Enabled: true
+  VersionAdded: '0.75'
```
Comment (bbatsov) [inline config/default.yml]: Is this in the style guide? If so - let's add a link to the rule.

File: lib/rubocop/cop/lint/raise_exception.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Lint
+      # This cop checks for `raise` or `fail` statements which is
```
Comment (bbatsov) [inline lib/rubocop/cop/lint/raise_exception.rb]: which are raising

File: lib/rubocop/cop/lint/raise_exception.rb [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+      # raise `Exception` class.
+      #
+      # @example
+      #   # bad
+      #   raise Exception, 'Error message here'
+      #
+      #   # good
+      #   raise StandardError, 'Error message here'
+      class RaiseException < Cop
+        MSG = 'Avoid `%<method>s` the `Exception` class. ' \
```
Comment (bbatsov) [inline lib/rubocop/cop/lint/raise_exception.rb]: I'd just use "Use `StandardError` over `Exception`." as the message here.

File: config/default.yml (hunk 3 above)
Comment (denys281) [reply config/default.yml]: no, we have only this one https://github.com/rubocop-hq/ruby-style-guide#no-blind-rescues

Comment (bbatsov) [reply config/default.yml]: Guess we can add something on the subject there then.

Comment (denys281) [reply config/default.yml]: @bbatsov Did you mean  in style guide?

Comment (bbatsov) [reply config/default.yml]: Yep.

Comment (denys281) [reply config/default.yml]: @bbatsov  https://github.com/rubocop-hq/ruby-style-guide/pull/775

**Comment by bbatsov** [comment]
Better late than never! :-) Thanks!

**Pull request merged**